dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,670 @@
+	// +optional
+	Count *CountDetails `json:"count,omitempty"`
 
-	// Device claims may be satisfied by choosing multiple devices instead
-	// of just a single device.
+	// Requirements describe additional contraints that all must be met by a device
+	// to satisfy the request.
+	//
+	// +optional
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
+
+// CountDetails defines how many instances are desired.
+type CountDetails struct {
+	// Minimum defines the lower limit. At least this many instances
+	// must be available (x >= minimum). The default if unset is one.
+	Minimum *int `json:"minimum"`
 
-	// MaxDevices allows the user to specify a maximum number of devices
-	// that may be allocated to sastisfy this claim. The default is equal
-	// to requests.devices. Thus, without specifying MaxDevices, the user
-	// will get exactly the number of devices specified in
-	// requests.devices.
+	// Maximum defines the upper limit. At most this many instances
+	// may be allocated (x <= maximum). The default if unset is unlimited.
+	Maximum *int `json:"maximum"`
+}
+
+// MatchModel must have one and only one field set.
+type MatchModel struct {
+	// All devices must have this attribute and its value must be the same.
+	//
+	// For example, if you specified "numa.dra.example.com" (a hypothetical example!),
+	// then only devices in the same NUMA node will be chosen.
 	//
 	// +optional
-	MaxDevices *int `json:"maxDevices,omitempty"`
+	// +listType=atomic
+	Attribute *string `json:"attribute,omitempty"`
 
-	// Requests allows the user to specify the minimum resource
-	// requirements that must be satisfied across all devices allocated for
-	// this claim.
+	// Future extension, not part of the current design:
+	// A CEL expression which compares different devices and returns
+	// true if they match.
 	//
-	// All drivers can the support "devices", resource, which represents
-	// the minimum number of devices to allocate.
+	// Because it would be part of a one-of, old schedulers will not
+	// accidentally ignore this additional, for them unknown match
+	// criteria.
 	//
-	// Other resource types are driver-specific. The values in `requests`
-	// represent the aggregate across all devices, not the per-device
-	// values.
+	// matcher string
+}
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the result of that was.
+type ResourceClaimStatus struct {
+	// ControllerName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started. It is empty when the claim was
+	// allocated through structured parameters,
 	//
-	// If not set, requests.devices will default to 1.
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
 	//
 	// +optional
-	Requests map[string]resource.Quantity `json:"requests,omitempty"`
+	ControllerName string `json:"controllerName"`
```

Comment:
Why is this needed?

Reply:
For deallocation when classic DRA is used. The controller name from the class might not be available anymore.

Speaking of "class" and "controller name": depending on the outcome of our class discussion, moving that field into the claim may be the better approach. Then classic DRA can be used without a class.